### ISSUE TITLE
Ability to execute a command on start and exit

### ISSUE THREAD

**Issue by SamSaffron** [krallin/tini#28]
I was wondering if it is in the scope of tini to allow for execution of a script on start and exit. 

Eg something like

```
tini --boot /etc/init.d/1/boot --shutdown /etc/init.d/3/shutdown my_process
```

This clearly can be done from "my_process" but I think it may be convenient to add these switches. 

Is this something you feel fits in tini or not?

**Comment by krallin** [comment]
Intuitively, this might be slightly out of scope for what I envision Tini being useful for, but I don't feel very strongly about it in the abstract. Did you have a specific use case in mind for this? This would help me form a more thoughtful opinion, I think!

Thanks! : )

**Comment by SamSaffron** [comment]
Yes, my exact use case is so I can put this little script to rest
https://github.com/discourse/discourse_docker/blob/master/image/base/boot

We like doing certain cleanup when booting, like cleaning up orphan socket
files or ensuring certain permissions on a shared volume

On teardown we like to stop services in a certain order or in certain
signals.

Totally understand if you feel it is out of scope, but I would be super
happy if it was added :)

**Comment by krallin** [comment]
I see; thank you for sharing! I have a bit on my plate, but I'll try to have a look this week and at least let you know where I stand : )

Cheers,

**Comment by SamSaffron** [comment]
No probs , thanks for tini :)

**Comment by krallin** [comment]
Hey @SamSaffron, I have an initial implementation in the https://github.com/krallin/tini/compare/pre-post-commands branch. If you'd like to play around with it a little bit, feel free : ) (this would be a big help!). Compile instructions are [in the README][0].

The arguments are `--pre` and `--post`. Both accept one and only argument (as you'd probably expect). 

Beyond that, I think the only gotcha is that the exit code for the `--post` command will be discarded in favor of the exit code from the main program (whereas if `--pre` fails, its exit code is returned and the main program is never spawned). 

I'm not sure this is the right way to do it, though. I'd be curious to hear your thoughts.

---

I still have to add tests and docs, and I'll play around a little bit with this before I make a decision to merge it in or not. There's a lot more code that I could reuse than I expected, though (the commit is pretty big, but most of it is refactoring), which I think is a good sign! : )

Cheers,

  [0]: https://github.com/krallin/tini#building-tini

**Comment by SamSaffron** [comment]
This is super awesome, I am going to try it out on Discourse docker and let you know how I go. 

The decisions you made there about exit codes are absolutely sound, my slight tweak would be:

- If post and main have a non zero exit code, return main exit code, log post exit code to STDERR
- If only post has non zero exit code, return it
- If only main has non zero exit code, return it 

Thanks so much for trying this out!!!

**Comment by krallin** [comment]
@SamSaffron thanks! Let me know how it goes :).

Your proposed tweaks make sense to me. I was actually considering just returning `main exit code + post exit code` (which is *probably* identical as far as anyone would care), but I like your proposal just as much.

One other note: the post command command runs even if the pre command one fails. I'm assuming it'd be used for cleanup mainly, so I *think* that's a sound approach, but me know if your perspective differs.

---

In other, clearer words:

- Pre and post always run if provided.
- Main only runs if pre didn't fail.

Cheers,

**Comment by SamSaffron** [comment]
Yes I think that is correct, you may want to clean up stuff you setup in pre, makes total sense.

**Comment by SamSaffron** [comment]
Having been debugging and integrating this stuff I came across a bit of a snag

If the order is

- pre
- process
- post

It means that post has no way of controlling some of the cleanup. 

If the order is 

- pre
- tini gets TERM
- post
- process gets TERM 

You can control cleanup process better. 

Concrete example tini launches a process with 2 sub processed, say PG and syslogd. You may want it to stop PG first and then syslogd second (so you don't lose logs), but if post runs after process is terminated (especially with -g) termination order is inconsistent. 

I am not sure this can and should be resolved by tini, but its worth thinking about. 

I guess an initial, fundamental question is ... should sending TERM to tini **guarantee** that child/children terminate or not? (sending a QUIT 10 seconds later or something) Cause whatever is decided strongly depends on that.

**Comment by krallin** [comment]
Inline,

> Having been debugging and integrating this stuff I came across a bit of a snag
> 
> If the order is
> 
> - pre
> - process
> - post
> 
> It means that post has no way of controlling some of the cleanup. 
>
> If the order is 
>
> - pre
> - tini gets TERM
> - post
> - process gets TERM 
> 
> You can control cleanup process better.

For better or worse, Tini doesn't actually know anything about the signals it forwards. In other words, SIGTERM isn't treated differently from any other signal. I'd rather keep this as-is if possible to make Tini as predictable as possible.

> Concrete example tini launches a process with 2 sub processed, say PG and syslogd. You may want it to stop PG first and then syslogd second (so you don't lose logs), but if post runs after process is terminated (especially with -g) termination order is inconsistent.

In this particular case, I think you might want to just run syslogd in its own process group. 

Here are a few options I can think of:

+ Spawn syslogd using the `--pre` script! Tini launches pre, main, and post in their own individual process group.
+ If you're not using Tini to launch Syslog (e.g. you're using a shell script or a process manager), then don't spawn syslogd directly. Instead, spawn something that first creates its own process group and *then* execs syslogd ([this program *seems* to do exactly that](https://github.com/maandree/setpgrp/blob/master/src/setpgrp.c))
+ And if you're lucky, maybe the process manager you're using (if any) has that option to begin with : ).

> I am not sure this can and should be resolved by tini, but its worth thinking about.

> I guess an initial, fundamental question is ... should sending TERM to tini guarantee that child/children terminate or not? (sending a QUIT 10 seconds later or something) Cause whatever is decided strongly depends on that.

As you've probably guessed at this point, my current thinking is that no, it shouldn't. 

As noted above, the main reason for this is to keep Tini as predictable as possible: it forwards signals, but it doesn't try to do anything smart about them.

Ideally, this helps keep it composable with other systems (e.g. Tini forwards the signals, and you can use a process manager like supervisord to do something smart about them). Besides, Tini is mostly used in Docker, and `docker stop` does the "term then kill" thing for users that do need it.

--

Curious to hear your thoughts, and thanks again for taking the time to answer : ). And of course, just let me know if I've misunderstood something about your use case!

Cheers,

**Comment by yosifkit** [comment]
This feels out of the scope of the "simplest init you could think of."  I see `tini` as a zombie reaper and signal forwarder, and am not sure how this is an extension that fits in that role.

As you show, this can easily be done by your own script, and probably needs to handle different signals in different ways so that only a `TERM` signal starts a "shutdown" process before sending the `TERM` to the main process.

This implementation also breaks the "spawn a single child".  With `--pre` I could run a process that forks and the parent terminates (`sh -c 'long-process &'`); that process would still be running when `tini` starts the main process.

**Comment by krallin** [comment]
Hey @yosifkit , thanks for your feedback : )

Inline,

> This feels out of the scope of the "simplest init you could think of."  I see `tini` as a zombie reaper and signal forwarder, and am not sure how this is an extension that fits in that role.

Just to be clear: are you talking about the entire feature request, or specifically about special-casing SIGTERM?

If you're referring to the entire feature request: see last paragraph.

If you're referring to special-casing SIGTERM: agreed.

> As you show, this can easily be done by your own script, and probably needs to handle different signals in different ways so that only a `TERM` signal starts a "shutdown" process before sending the `TERM` to the main process.

Same question here (though here I'm reasonably sure you're arguing against special-casing SIGTERM, and I agree with you)

>This implementation also breaks the "spawn a single child".  With `--pre` I could run a process that forks and the parent terminates (`sh -c 'long-process &'`); that process would still be running when `tini` starts the main process.

For what it's worth, you can already do this anyway with e.g. the following "main" process:

```
#!/bin/sh
sh -c 'sleep 20' &
exec sleep 5
```

So, there's little point in Tini preventing you from doing this. 

Now, you're entirely right: the `--pre` command isn't particularly useful. You can do the exact same thing with a script like the one above. Having `--pre` just makes it a bit more discoverable / obvious.

However I don't think you can reproduce `--post` so easily without support from Tini (or at least I haven't found a way to — but let me know if you see things differently!).

In other words, the real feature here is `--post`. However, considering the exact same logic is needed to support `--pre` and `--post`, I think it's worth having both for the sake of symmetry.

---

Thanks!

**Comment by danilobuerger** [comment]
I agree with @yosifkit 
It kind of feels out of scope. Maybe i don't understand this properly. But why cant you run tini and then the post command afterwards?

**Comment by krallin** [comment]
Hey @danilobuerger,

To run Tini and the post command afterwards (in particular in a container), you need another program running as PID 1 (because if Tini runs as PID 1, when Tini exits, there's nothing to run your post command).

So, your hierarchy looks like this:

```
something (pid 1) -> tini -> main
```

But, the `something` program running as PID 1 arguably needs to satisfy a few constraints:

- It needs to forward signals, otherwise signals won't get to Tini and won't get to your `main` program.
- It needs to reap zombies (though alternatively you *could* use Tini's subreaper mode here).
- It needs to `wait` on Tini, and presumably forward its exit code (so you can know if `main` failed).

So, basically, `something` needs to do just about everything Tini already does. So, why not have Tini *be* that `something`?

Let me know if that makes sense! I'm not ruling out that I might be missing something (but at the very least I don't think it's something obvious! ;) ).

Thanks!

**Comment by danilobuerger** [comment]
So we would have the same problem with your latest solution in #29 of not forwarding signals to tini, right?

**Comment by krallin** [comment]
I'm sorry @danilobuerger , I didn't understand your question here : (

**Comment by danilobuerger** [comment]
In #29 you advised to use the following as entrypoint:

```
#!/bin/sh
sh -c 'sleep 10 && echo "TIMEOUT" && kill -s TERM 1' &
exec tini -- sh -c 'sleep 20'
```

If i do that, tini doesnt receive the signals from docker, right?

**Comment by krallin** [comment]
@danilobuerger It does, actually, because `exec` *replaces* the program currently running (`/bin/sh`) with `tini`.

In other words, before `exec` (the last line), the process hierarchy looks like this:

```
sh (pid 1) -> sh -> sleep 10
```

And afterwards, it looks like this:

```
tini (pid 1) -> sh -> sleep 10
```

And then, once Tini spawns another shell and that shell runs `sleep 20`:

```
tini (pid 1) -> sh -> sleep 10
            \-> sh -> sleep 20
```

**Comment by danilobuerger** [comment]
Ah, thanks for your, as always, helpful explanation!!

**Comment by oblique** [comment]
I also believe that pre/post is not actually needed because it can be reproduced with a script. But if you are going to merge these features then tini should also have the main program as optional parameter if `--pre` is used.

A valid senarior is a container that has only daemons where you start them with `--pre`, so the main program must be a small script that has an infinite loop until it gets a TERM signal.

But again, isn't pre/post complicating things? Can you give a scenario where `--post` can not reproduced by a small script?

**Comment by krallin** [comment]
Thanks for your feedback @oblique,

Just to be clear; the question isn't really whether a given  `--post` command can be run without Tini using a shell script. 

Instead, it's whether you can implement a given `--post` command without writing a script that essentially amounts to re-implementing Tini

Indeed, bear in mind that you can re-implement 90% of what Tini does with a small shell script.  One of the main reasons for using Tini is that it's convenient: while you could re-implement *most* of what Tini does in a shell script, it's more work (and less readable) than just installing and using Tini.

--

The use case for supporting `--post` in Tini is essentially any case where you're running a co-process that you'd like to be terminated cleanly after the main process has exited (rather than scrapped by the kernel when PID 1 exits). The co-process would be `tail` sending logs to stdout because your main app doesn't support logging to stdout, or something a lot more involved.

Tini, by design, only cares about its immediate child, so if you start your co-process by using something like this as your target for Tini:

```
#!/bin/sh
my_coprocess &
exec my_process
```

Then, even if you use the `-g` flag, `my_coprocess` won't exit cleanly *if the immediate child exits first*.

If you want or need to kill it cleanly, then you can't really do that with Tini right now. Add `--post` is a wait to give users an easy way to do so if they actually care (note: currently the design of running a single command isn't ideal, I might end up simply passing the `--post` command to `sh -c`, which would make it possible e.g. inline `kill -s TERM $(pgrep my_coprocess)`).

Of course, just let me know if you think that doesn't make sense! :smile: 

-- Thomas

**Comment by oblique** [comment]
> while you could re-implement most of what Tini does in a shell script, it's more work (and less readable) than just installing and using Tini.

Ok this is a valid point. So what about this script?

```
#!/bin/bash

EXITING=0

set_exit_flag() {
    EXITING=1
}

trap set_exit_flag SIGINT SIGTERM

/etc/init.d/ssh start
/etc/init.d/nginx start

while [[ $EXITING -eq 0 ]]; do
    sleep 1
done

/etc/init.d/nginx stop
/etc/init.d/ssh stop
```

As I said before, shouldn't the immediate child be optional if `--pre` and `--post` is set? Because with the pre/post I need to have the following:

**pre:**
```
#!/bin/sh
/etc/init.d/ssh start
/etc/init.d/nginx start
```

**main:**
```
#!/bin/bash

EXITING=0

set_exit_flag() {
    EXITING=1
}

trap set_exit_flag SIGINT SIGTERM

while [[ $EXITING -eq 0 ]]; do
    sleep 1
done

exit 0
```

**post:**
```
#!/bin/sh
/etc/init.d/nginx stop
/etc/init.d/ssh stop
```

**Comment by krallin** [comment]
> > while you could re-implement most of what Tini does in a shell script, it's more work (and less readable) than just installing and using Tini.
> 
> Ok this is a valid point. So what about this script?
> 
> ```
> #!/bin/bash
> 
> EXITING=0
> 
> set_exit_flag() {
>     EXITING=1
> }
> 
> trap set_exit_flag SIGINT SIGTERM
> 
> /etc/init.d/ssh start
> /etc/init.d/nginx start
> 
> while [[ $EXITING -eq 0 ]]; do
>     sleep 1
> done
> 
> /etc/init.d/nginx stop
> /etc/init.d/ssh stop
> ```

I'm sorry, I have no idea of what you're trying to demonstrate with this script. 

If you'd like to clarify, I'm all ears.

```
> As I said before, shouldn't the immediate child be optional if --pre and --post is set? Because with the pre/post I need to have the following:
> 
> pre:
> 
> #!/bin/sh
> /etc/init.d/ssh start
> /etc/init.d/nginx start
> main:
> 
> #!/bin/bash
> 
> EXITING=0
> 
> set_exit_flag() {
>     EXITING=1
> }
> 
> trap set_exit_flag SIGINT SIGTERM
> 
> while [[ $EXITING -eq 0 ]]; do
>     sleep 1
> done
> 
> exit 0
> post:
> 
> #!/bin/sh
> /etc/init.d/nginx stop
> /etc/init.d/ssh stop
```

At this point you're essentially using Tini as a process manager. If you need a process manager, I'd recommend using Tini to spawn a  process manager (e.g. Supervisor).

Besides, making the main child optional and defaulting to a sleep loop (which I guess is what you're suggesting here?) would make Tini extremely confusing. I'd rather avoid that.

**Comment by oblique** [comment]
I use the script from above to start a container that has only sshd and httpd. But because the immediate child should be alive, I have a loop until it receives SIGTERM from tini and I proceed for cleanup.

Ok, I will take a look at supervisor.

**Comment by krallin** [comment]
Thanks for clarifying @oblique. Just to be clear: if I add those commands, anything you're doing today will continue working just fine (you don't *have* to use them: they'll be just an option).

**Comment by javabean** [comment]
You may want to have a look at [ContainerPilot](https://github.com/joyent/containerpilot) for some interesting ideas; in particular the `preStart`, `preStop` and `postStop`steps which I find more complete than a simple `pre` + `post` proposition.
From the [documentation](https://www.joyent.com/containerpilot/docs/start-stop):

>`preStart`
>The executable (and its arguments) that will be called immediately **prior** to starting the shimmed application. This field is optional. If the `preStart` handler returns a non-zero exit code, ContainerPilot will exit.
>
>`preStop`
>The executable (and its arguments) that will be called immediately **before** ContainerPilot stops the shimmed application. This happens when ContainerPilot receives a `SIGTERM` signal, such as from Docker during a `docker stop …`. This field is optional. ContainerPilot will wait until the `preStop` handler exits before terminating the shimmed application. Note that the Docker Engine will send a `SIGKILL` after the timeout on `docker stop` expires, so you'll want to configure that timeout to allow enough time for your `preStop` to expire.
>
>`postStop`
>The executable (and its arguments) that will be called immediately **after** the shimmed application exits. This field is optional. If the `postStop` handler returns a non-zero exit code, ContainerPilot will exit with this code rather than the application's exit code.

I do agree that the `preStart` is totally redundant with a simple start.sh script, but it is there for the sake of completeness.

On a side note, `preStop` would probably require tini to handle `SIGTERM` as a special case (+ what happens when one configures another signal on `docker stop`? We should have a special `--stop-signal` like Docker for this purpose).
With that in mind, I would be quite happy to only have a single `postStop` handler, and keep tini's code lean and clean! :-)

**Comment by ricardobranco777** [comment]
For this reason I still use [minit](https://github.com/chazomaticus/minit).  

Another useful feature would be to unsetenv() environment variables after the startup script runs, because of the bad practice of using environment variables for passing passwords to containers.

**Comment by krallin** [comment]
Thanks for the link to Minit @ricardobranco777 — I think it's fine and even desirable to have different init tools for different use cases (and it's good to have a link to it for folks that land on this issue). Tini seems to fit the "minimal init for single-process containers" niche pretty well, and I think it's good that Minit fits the "more complete system" one.

Of course, it's also possible to use Tini to launch e.g. `supervisord` or a similar tool if one wants a more complete system and Tini as PID 1, but I think Tini doesn't need to be the init system to rule them all :)

---

As for `unsetenv`, I'd suggest using Tini to launch [`env`](https://linux.die.net/man/1/env) and then launch your command; i.e.:

```
tini env -u MY_PASSWORD -u SOME_OTHER_VAR my_program
```

(and then if you need to change users, you could use `gosu`, etc.!)

... minimal composable tools, [as they say](https://en.wikipedia.org/wiki/Unix_philosophy) 😄 

---

I should probably have closed this issue some time ago; I'll do so now.
